### PR TITLE
merge im2deep into timsrescore branch 

### PR DESCRIPTION
**Pull request by ArthurDeclercq**
- Update basecontainer of docker file
- add IM2Deep to MS²Rescore
- add IM2Deep to the MS²Rescore GUI
- add IM2Deep correlation plot to the report
- Switched the pyteomics MGF and MZML to indexed MGF and MZML class for faster parsing of IM and RT from spectrum files

### CODE REVIEW

File: ms2rescore/parse_spectra.py [hunk 1]
```diff
@@ -26,122 +26,110 @@ def get_missing_values(config, psm_list, missing_rt=False, missing_im=False):
...
 
-            for value_dict, value in zip([rt_dict, im_dict], ["retention_time", "ion_mobility"]):
-                if value_dict:
-                    try:
-                        psm_list_run[value] = [value_dict[psm.spectrum_id] for psm in psm_list_run]
-                    except KeyError:
-                        raise ParsingError(
-                            f"Could not parse {value} values from spectrum file for run {run}."
-                        )
-
```
Comment (paretje) [inline ms2rescore/parse_spectra.py:55]: I'm not sure about the name of this method, `get_missing_values`, given that it doesn't return missing values, but actually fills them with data from the mgf or mzml.

File: ms2rescore/report/generate.py [hunk 2]
```diff
@@ -338,6 +338,29 @@ def _get_features_context(
             }
         )
 
+    # IM2Deep specific charts
+    if "im2deep" in feature_names:
+        import deeplc.plot
+
+        # TODO: the axis titles are gonna be retention time based so need to make them changeable in deeplc.plot
```
Comment (paretje) [inline ms2rescore/report/generate.py:345]: Should this be resolved in DeepLC before merging this?

File: ms2rescore/feature_generators/im2deep.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    def __init__(
+        self,
+        *args,
+        lower_score_is_better: bool = False,
+        spectrum_path: Optional[str] = None,
+        processes: int = 1,
+        calibrate_per_charge: bool = True,
+        **kwargs,
+    ):
+        """Placeholder"""  # TODO
```
Comment (paretje) [inline ms2rescore/feature_generators/im2deep.py:34]: Reminder ;)

File: ms2rescore/feature_generators/im2deep.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        t_diff : float, optional
+            Factor to convert Celsius to Kelvin, by default 273.15"""
+
+        SUMMARY_CONSTANT = 18509.8632163405
+        reduced_mass = (mz * charge * mass_gas) / (mz * charge + mass_gas)
+        return (SUMMARY_CONSTANT * charge) / (
+            np.sqrt(reduced_mass * (temp + t_diff)) * 1 / reverse_im
+        )
+
+    # TODO: Make threshold a parameter
```
Comment (paretje) [inline ms2rescore/feature_generators/im2deep.py:170]:
```suggestion
```
Or do you mean expose it as I see `add_features` doesn't have the parameter.

**Review by paretje — CHANGES_REQUESTED** [review]

File: ms2rescore/feature_generators/im2deep.py [hunk 5]
```diff
@@ -1,5 +1,11 @@
 """
-Placeholder
+IM2Deep ion mobility-based feature generator.
+
+IM2Deep is a fully modification-aware peptide ion mobility predictor. It uses a deep convolutional
+neural network to predict retention times based on the atomic composition of the (modified) amino
+acid residues in the peptide. See
+`github.com/compomics/deeplc <https://github.com/compomics/IM2Deep>`_ for more information.
```
Comment (RalfG) [inline ms2rescore/feature_generators/im2deep.py:7]: Todo: Replace DeepLC with im2deep

**Review by RalfG — CHANGES_REQUESTED** [review]

File: ms2rescore/feature_generators/im2deep.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
+"""
+IM2Deep ion mobility-based feature generator.
+
+IM2Deep is a fully modification-aware peptide ion mobility predictor. It uses a deep convolutional
+neural network to predict retention times based on the atomic composition of the (modified) amino
+acid residues in the peptide. See
+`github.com/compomics/deeplc <https://github.com/compomics/IM2Deep>`_ for more information.
```
Comment (RalfG) [inline ms2rescore/feature_generators/im2deep.py:7]:
```suggestion
`github.com/compomics/IM2Deep <https://github.com/compomics/IM2Deep>`_ for more information.
```

**Review by RalfG — APPROVED** [review]

**Pull request merged**